addons/io_scene_gltf2/blender/exp/gltf2_blender_image.py

```diff
@@ -17,118 +17,195 @@
-
-    @property
-    def height(self):
-        return self.shape[1]
-
-    @property
-    def channels(self):
-        return self.shape[2]
+class ExportImage:
+    """Custom image class.
+
+    An image is represented declaratively by giving the data that should fill
+    its red, green, blue, and alpha channels. For example:
+
+        self.fills = {
+            Channel.R: Attachment(image=bpy.data.images['Im1'], src_chan=Channel.B),
+            Channel.G: White(),
+        }
+
+    This says that the ExportImage's R channel should be filled with the B
+    channel of the Blender image 'Im1', and the ExportImage's G channel should
+    be filled with all 1.0s. Undefined channels mean we don't care what values
+    that channel has.
+
+    This is flexible enough to handle the case where eg. the user used the R
+    channel of one image as the metallic value and the G channel of another
+    image as the roughness, and we need to synthesize an ExportImage that packs
+    those into the G and B channels for glTF.
+
+    When we need to encode the ExportImage, cases like that require us to do
+    pixel manipulation to put the individual channels together, which is slow
+    and memory-hungry, but something like the following
+
+        self.fills = {
+            Channel.R: Attachment(image=im, src_chan=Channel.R),
+            Channel.G: Attachment(image=im, src_chan=Channel.G),
+            Channel.B: Attachment(image=im, src_chan=Channel.B),
+        }
+
+    has a happy path: we can just ask Blender to save the image im and read the
+    bytes back, which is much faster.
+    """
+    # FUTURE_WORK: as a method to allow the node graph to be better supported, we could model some of
+    # the node graph elements with numpy functions
 
-    def __getitem__(self, key):
-        """returns a new ExportImage with only the selected channels"""
-        return ExportImage(self._img[:, :, key])
+    def __init__(self):
+        self.fills = {}
 
-    def __setitem__(self, key, value):
-        """set the selected channels to a new value"""
-        if isinstance(key, slice):
-            self._img[:, :, key] = value.img
-        else:
-            self._img[:, :, key] = value.img[:, :, 0]
+    @staticmethod
+    def from_blender_image(image: bpy.types.Image):
+        export_image = ExportImage()
+        for chan in range(image.channels):
+            export_image.attach(chan, Attachment(image=image, src_chan=chan))
+        return export_image
 
-    def append(self, other):
-        if self.channels + other.channels > self._max_channels:
-            raise RuntimeError("Cannot append image data to this image "
-                               "because the maximum number of channels is exceeded.")
+    def attach(self, chan: Channel, attachment: Attachment):
+        self.fills[chan] = attachment
 
-        self._img = np.concatenate([self.img, other.img], axis=2)
+    def fill_white(self, chan: Channel):
+        self.fills[chan] = White()
```

Comment:
This is minor, but the naming of "attach" and "Attachment" don't feel very clear to me. On the other hand, `fill_white(Channel.B)` is quite easy to understand. Since the Attachment class is only used once outside this class, maybe replace it with a similar function that accepts an image?

```
image.fill_image(image, Channel.R)
image.fill_white(Channel.G)
```

If the container class is useful regardless, maybe names like ImageFill or SourceImage would be more obvious?

Reply:
I'd like to improve the names too. `Attachment` was loosely based on color attachments for a render pass.

How about this then?

````py
# Outside this file
export_image.fill_image(image, dst_chan=dst_chan, src_chan=src_chan)
export_image.fill_white(dst_chan=dst_chan)

# Only used inside this file
class FillImage # fills with an image channel
class FillWhite # fills with all 1.0s
````

Reply:
Sounds good to me!

Reply:
Alright, done. I also renamed `has_channel`->`is_filled`.